server/src/main/java/com/defold/extender/services/UserUpdateService.java

```diff
@@ -0,0 +1,81 @@
+package com.defold.extender.services;
+
+import org.springframework.stereotype.Service;
+import org.springframework.security.provisioning.UserDetailsManager;
+import org.springframework.security.provisioning.InMemoryUserDetailsManager;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.core.userdetails.User;
+import org.springframework.security.core.userdetails.User.UserBuilder;
+import org.springframework.core.io.Resource;
+import java.util.Properties;
+import java.util.Arrays;
+import java.util.List;
+import java.io.IOException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+@Service
+public class UserUpdateService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserUpdateService.class);
+
+    @Autowired
+    private InMemoryUserDetailsManager userDetailsManager;
+
+    @Value("${extender.authentication.users}")
+    Resource usersResource;
+
+    @Value("${extender.authentication.update-interval}")
+    long updateInterval;
+
+    private long lastUpdateTimestamp = 0;
+
+    private Properties loadUsers() {
+        Properties users = new Properties();
+        try {
+            users.load(usersResource.getInputStream());
+        }
+        catch(IOException e) {
+            LOGGER.info("UserUpdateService - Unable to update users. " + e.getMessage());
+        }
+        return users;
+    }
+
+    // user = password,ROLE_ANDROID,ROLE_IOS,ROLE_HTML5,ROLE_WINDOWS,ROLE_LINUX,ROLE_MACOS,ROLE_SWITCH,disabled
```

Comment:
Is this a "standard" format?
If we later need to add/remove info to this (not sure what, but let's say email), will it be easy to do? Will it be easy to remove fields?
Not sure we need to, just making sure the format is as solid we can make it/want it.

Reply:
Yeah, so the InMemoryUserDetailsManager accepts a Properties instance as one of its constructors:

https://docs.spring.io/spring-security/site/docs/4.2.20.RELEASE/apidocs/org/springframework/security/provisioning/InMemoryUserDetailsManager.html#InMemoryUserDetailsManager-java.util.Properties-

I can't seem to find the example of how it is used anymore.

BUT since we aren't creating the InMemoryUserDetailsManager with a Properties instance in the constructor we're actually free to change the format if we like.

Reply:
👍 
It's good that we can change it.

Workflow question: Will it be easy?
I'm thinking of the update workflow here.
Let's say we update the parsing function, since we want to add/remove a field.
We then need to build stage server, and point it to a separate users.txt.
During these two weeks (the beta period), we need to update two separate users.txt to keep them working for our users.
Then, after two weeks, we can merge the stage/users.txt to the prod/users.txt.

I'm not going to push for any changes at this point though.since this is yet theoretical and not an actual use case yet.

Reply:
Yeah, I get what you are saying. What we do know is that password comes first, then a list of authorities/roles all prefixed with ROLE_ and finally a value indicating if the user is enabled or not. We're not locked into a format which we cannot change at this moment, but I agree that it makes sense to pick a more extensible format.